### PR TITLE
Move refactor

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Make move an inline proc, no longer 5 proc call for a single diagonal movement.

Tested so far:
 - You cannot phase between two walls while moving diagonally
 - You cannot phase between two mobs if you weren't able to bump/shuffle either
 - You cannot phase throught baricades or windows (true this time)
 - Pulling seems to have normal behaviour
 - Push seems to have normal behaviour (you can now push diagonally as a bonus)
 - Shuffling seems to have normal behaviour (You can now shuffle diagonally as a bonus)
 - Slipping seems to work normally
 - Projectiles works normally
 - Space walking seems to be normal
 - Throwing works normally

Benchmark (the time is a lie, it's for 30 seconds)

Old code 
image

New code: 
image

So according to this benchmark (conditions are roughly similar, called this at the same moment on the same map at the same place on a carbon human, for 30 secs), this is a 14% perf increase for cardinal moves and a 23% for diagonal move

Here's the benchmark code used :
image

Probably a change of behaviour in how the after image for bubble gum works. (Only once per diag move rahter than twice)

## Why It's Good For The Game

Probably better perf, as we are proc calling less. I'll get profile data from tgmc if this passes review process.

Also cool diag move : 

https://user-images.githubusercontent.com/25491240/141023940-615bfc62-2072-453c-bccc-b9811182c060.mp4

Move() is the most expensive non yielding code on tg and tgmc, even the slightest perf increase can help.

## Changelog

:cl:
code: Probably better performance with the Move() proc. 14% perf increase for cardinal moves and 23% for diagonal move
qol: You can shuffle/push diagonally
/:cl:

### CODE REVIEW

File: code/game/atoms_movable.dm
Comment (BraveMole) [inline code/game/atoms_movable.dm:477]: current_turf was initialised like this : var/turf/current_turf = loc.

There was no reason to believe loc was a turf, nor the cast needed. I assume it was shitcode, and oldloc works just fine

File: code/game/atoms_movable.dm
Comment (BraveMole) [inline code/game/atoms_movable.dm:391]: i'm not really proud of this, but i haven't thought of another solution. Suggestion are welcomed

File: code/game/atoms_movable.dm
Comment (BraveMole) [inline code/game/atoms_movable.dm:382]: this loc == newloc check didn't return altogether.

That might be a behaviour change, althought i don't know when would Move be called with loc == new loc and what it should do in that case.

Insight from someone more experienced would be appreciated

File: code/game/atoms_movable.dm
Comment (Kylerace) [inline code/game/atoms_movable.dm:390]: Use NONE instead of 0 here. Also I'm pretty sure you can do this more efficiently by storing which two orthogonal directions the movement direction is composed of and then checking the north/south one then the east/west one in this manner. Should be profiled though

**Review by Kylerace — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
This strikes me as a bit of a bad idea, since either you duplicate the logic of Move() in Entered() and get the same numbers, or you don't and break things.

I'm not sure what you mean by shuffling, but pushing works diagonally currently

**Comment by BraveMole** [comment]
> This strikes me as a bit of a bad idea, since either you duplicate the logic of Move() in Entered() and get the same numbers, or you don't and break things.

There's a bunch of stuff you should be doing only once in a diag move.

Calling moved, checking the loc, setting the dir, checking pulling, checking buckled, setting the glide speed, the area check,  checking if you can leave the current turf, etc.

On top of that, inlining this proc prevent Move being called 5 times for a single movement, which creates proc overhead.

> I'm not sure what you mean by shuffling, but pushing works diagonally currently

shuffling = changing position with something.

image

And pretty sure you can't push diagonnaly, at least you can't on tgmc and the code is the same

**Comment by LemonInTheDark** [comment]
> There's a bunch of stuff you should be doing only once in a diag move.
> 
> Calling moved, checking the loc, setting the dir, checking pulling, checking buckled, setting the glide speed, the area check, checking if you can leave the current turf, etc.
> 
> On top of that, inlining this proc prevent Move being called 5 times for a single movement, which creates proc overhead.

In-lining is fine, I mean for the things Move() calls. Calling Moved() should absolutely happen for each tile moved, the same as calling Move().
There's a lot of code that relies on those procs being called for every tile moved. Whether they should or not is another question, but they do currently, which I assume is why you're seeing any major benefits from this, since Move() itself is very minor in cost compared to the things it calls. That's not to say inlining is a bad thing, but dropping the second move seems shortsighted
> > I'm not sure what you mean by shuffling, but pushing works diagonally currently
> 
> shuffling = changing position with something.
> 
> image
> 
> And pretty sure you can't push diagonally, at least you can't on tgmc and the code is the same
Ah I see, thank you.
You can push diagonally. It's a Move(push_target, get_dir(pusher, push_target)), and diagonal moves works just fine

**Comment by BraveMole** [comment]
I fail to see what item need to call Moved twice in a diagonal move? 
image
Possibly this one maybe

**Comment by LemonInTheDark** [comment]
Projectiles and humans seem to be causing problems, I assume for similar reasons

> 
> 
> I fail to see what item need to call Moved twice in a diagonal move? image Possibly this one, but shouldn't this behaviour be on bump anyway

There's a large amount of things that expect Move() or Moved(). I don't see the point of moving them all to entered since you just end up in the same position but again

**Comment by ShizCalev** [comment]
> 
> 
> I fail to see what item need to call Moved twice in a diagonal move? image Possibly this one, but shouldn't this behaviour be on bump anyway

Hi, I rewrote most of the mule code a while back. I did attempt to move that over to Bump when I did my rewrite - iirc Bump() is called on the first dense thing in contents - ergo if you're laying down, you won't get run over.

**Comment by TimeIsEndless** [comment]
Two Issues so far. When moving diagonally you can squeeze through directional windows. and you can also enter tiles containing other players while moving diagonally.

**Comment by jlsnow301** [comment]
Can throw objects through walls (or at least onto their tile)
Imgur img

File: code/game/atoms_movable.dm
Comment (Rohesie) [inline code/game/atoms_movable.dm:407]:
```suggestion
		if((direction & NORTH) && get_step(loc, NORTH)?.Enter(src))
			can_pass_diagonally = NORTH
		else if((direction & EAST) && get_step(loc, EAST)?.Enter(src))
			can_pass_diagonally =  EAST
		else if((direction & WEST) && get_step(loc, WEST)?.Enter(src))
			can_pass_diagonally = WEST
		else if((direction & SOUTH) && get_step(loc, SOUTH)?.Enter(src))
			can_pass_diagonally = SOUTH
		else
			moving_diagonally = FALSE
			return
		moving_diagonally = FALSE
```
Made this code a tiny bit more succinct.

Comment (BraveMole) [reply code/game/atoms_movable.dm:407]: yeah, much better. Thanks

File: code/game/atoms_movable.dm
Comment (LemonInTheDark) [inline code/game/atoms_movable.dm:434]: This is silly and I won't abide it

File: code/game/atoms_movable.dm
Comment (BraveMole) [inline code/game/atoms_movable.dm:7]: @Kylerace Tell me what i have to check for this.

At first glance, look uneeded and nothing went wrong in testing

File: code/game/atoms_movable.dm
Comment (BraveMole) [reply code/game/atoms_movable.dm:430]: Discussed on disc

File: code/game/atoms_movable.dm
Comment (Kylerace) [reply code/game/atoms_movable.dm:7]: its hard to explain succinctly but moving something with connect_loc_behalf attached to it like slippery items (like banana peels) and doing nested movement calls on them such that they move from A to B to C and Moved(old_loc = B) is called before Moved(old_loc = A) used to reproduce the issue but shouldnt since c_l_b is a component now. i think what i used to reproduce it was using a conveyor belt to move a banana peel onto a hyperspace turf which moves it to another loc, and that should cause a runtime if htis is still a problem

**Comment by optimumtact** [comment]
What is the status of this PR, and how likely is it to be ready for merge soon.

**Comment by BraveMole** [comment]
> What is the status of this PR, and how likely is it to be ready for merge soon.

Still waiting for a review.

I think it's feature wise the same as old move code (maybe one or two exception, but i haven't caught anything in code or in testing), and roughly 10% more performant.
This is no longer allowing to push diagonally or shuffle diagonnaly, since i had to redo it multiple time based on maintainer request who wanted to make it totally identical behaviour wise. 

This version wasn't tested neither or tgmc or tg, the one before was and it had one bug with connect loc not being cleaned (and that bug has been fixed)

**Comment by BraveMole** [comment]
This contains bug that are not fixeable if i want to keep complying with maintainers request.

**Pull request closed without merging**
